File: archive/r/ruby/bubble-sort.rb
def bubble_sort(numbers)
  arr = numbers.dup
  n = arr.length

  (n - 1).times do |i|
    (n - i - 1).times do |j|
      next unless arr[j] > arr[j + 1]

      arr[j], arr[j + 1] = arr[j + 1], arr[j]
    end
  end

  arr
end

def parse_input
  raw = ARGV.first
  raise ArgumentError unless raw

  numbers = raw.split(",").map { Integer(it.strip, exception: false) }
  raise ArgumentError if numbers.any?(nil) || numbers.length < 2

  numbers
end

def usage
  warn %(Usage: please provide a list of at least two integers to sort in the format "1, 2, 3, 4, 5")
end

begin
  numbers = parse_input
  puts bubble_sort(numbers).join(", ")
rescue ArgumentError
  usage
end

File: archive/r/rust/bubble-sort.rs
use std::env;
use std::process;

fn bubble_sort(arr: &mut Vec<i32>) {
  let mut swap_occured = true;

  while swap_occured {
    swap_occured = false;

    for i in 1..arr.len() {
      if arr[i - 1] > arr[i] {
        arr.swap(i - 1, i);
        swap_occured = true;
      }
    }
  }
}

fn parse_int(s: &String) -> Option<i32> {
  match s.trim().parse::<i32>() {
      Ok(n) => Some(n),
      Err(e) => None,
  }
}

fn parse_int_list(s_list: &String) -> Vec<i32> {
  let results: Vec<Option<i32>> = s_list.split(",")
    .map(|s| parse_int(&s.to_string()))
    .collect();
  if results.iter().any(|s| s.is_none()) {
    vec![]
  }
  else {
    results.iter().map(|result| result.unwrap()).collect()
  }
}

fn main() {
  let err_msg = "Usage: please provide a list of at least two integers to sort in the format \"1, 2, 3, 4, 5\"";
  let args: Vec<String> = env::args().collect();

  if args.len() < 2 {
    println!("{err_msg}");
    process::exit(0);
  }

  let mut arr: Vec<i32> = parse_int_list(&args[1]);

  if arr.len() < 2 {
    println!("{err_msg}");
    process::exit(0);
  }

  bubble_sort(&mut arr);
  println!("{:?}", arr);
}

